include/deal.II/matrix_free/dof_info.templates.h

```diff
@@ -170,120 +177,126 @@ namespace internal
 
-      unsigned int dofs_this_cell = (cell_active_fe_index.empty()) ?
-                                    dofs_per_cell[0] : dofs_per_cell[cell_active_fe_index[cell_number]];
-      for (unsigned int i=0; i<dofs_this_cell; i++)
+      Assert(dofs_per_cell.size() == 1 || cell_number < cell_active_fe_index.size(),
+             ExcInternalError());
+      const unsigned int fe_index = dofs_per_cell.size() == 1 ? 0 : cell_active_fe_index[cell_number];
+      unsigned int dofs_this_cell = dofs_per_cell[fe_index];
+      const unsigned int n_components = start_components.back();
+      for (unsigned int comp=0; comp<n_components; ++comp)
         {
-          types::global_dof_index current_dof =
-            local_indices[lexicographic_inv[i]];
-          const std::vector<std::pair<types::global_dof_index,double> >
-          *entries_ptr =
-            constraints.get_constraint_entries(current_dof);
-
-          // dof is constrained
-          if (entries_ptr != nullptr)
+          std::pair<unsigned short,unsigned short> constraint_iterator (0,0);
+          for (unsigned int i=component_dof_indices_offset[fe_index][comp];
+               i<component_dof_indices_offset[fe_index][comp+1]; i++)
             {
-              // in case we want to access plain indices, we need to know
-              // about the location of constrained indices as well (all the
-              // other indices are collected by the cases below)
-              if (current_dof < first_owned || current_dof >= last_owned)
-                {
-                  ghost_dofs.push_back (current_dof);
-                  cell_at_boundary = true;
-                }
-
-              // check whether this dof is identity constrained to another
-              // dof. then we can simply insert that dof and there is no need
-              // to actually resolve the constraint entries
+              types::global_dof_index current_dof =
+                local_indices[lexicographic_inv[i]];
               const std::vector<std::pair<types::global_dof_index,double> >
-              &entries = *entries_ptr;
-              const types::global_dof_index n_entries = entries.size();
-              if (n_entries == 1 && std::fabs(entries[0].second-1.)<1e-14)
+              *entries_ptr =
+                constraints.get_constraint_entries(current_dof);
+
+              // dof is constrained
+              if (entries_ptr != nullptr)
                 {
-                  current_dof = entries[0].first;
-                  goto no_constraint;
-                }
+                  // in case we want to access plain indices, we need to know
+                  // about the location of constrained indices as well (all the
+                  // other indices are collected by the cases below)
+                  if (current_dof < first_owned || current_dof >= last_owned)
+                    {
+                      ghost_dofs.push_back (current_dof);
+                      cell_at_boundary = true;
```

Comment:
maybe rename it to `cell_at_partition_boundary` as this is not the domain boundary, right?

Reply:
Yes, this is indeed a better name.

Reply:
Or let us choose `cell_at_subdomain_boundary` because we use those terms elsewhere.